src/grpc/stream_infer_handler.cc

```diff
@@ -142,9 +142,16 @@ ModelStreamInferHandler::Process(InferHandler::State* state, bool rpc_ok)
   // threads.
   {
     std::lock_guard<std::recursive_mutex> lock(state->step_mtx_);
-    if (state->IsGrpcContextCancelled()) {
-      bool resume = state->context_->HandleCancellation(state, rpc_ok, Name());
-      return resume;
+    if (state->context_->ReceivedNotification()) {
```

Comment:
Checking `ReceivedNotification()` is protected by `state->step_mtx_`, but setting `state->context_->SetReceivedNotification(true);` appears unprotected. Is this a race?

Reply:
There is no hazard in this case as there is a single streaming infer handler thread that receives the notification state and then later processes it.
The mutex protection is mainly to protect the state transitions happening within HandleCancellation()/HandleCompletion() against the ResponseComplete callback.

Reply:
Can you decrease the scope of the mutex to only the necessary section for future clarity then?